### DISCUSSION TITLE
Filter Work Items after the previous pipeline run or Get the Work Items for certain period of time

### DISCUSSION THREAD

**Discussion by khopithan** [rfennell/AzurePipelines#1358] [Q&A] [answered]
Hi @rfennell,

Thank you for your amazing work on Azure DevOps. I need some clarifications regarding Work Items gathering in the pipeline running.

When pipeline running, it will gather the Work Items that are associated with the commit/commits. Then, it will appear in the generated release note either the work item is closed or not.

After that, if we run the pipeline again, will it gather the old work items as well?

Can we get Work Items between specific time span? or the Work Items that are associated after the previous pipeline running?

Regards,
Khopi

**Comment by rfennell** [comment, marked as answer]
The WI & commits are returned based on on the Azure DevOps API calls to get changes between build runs 
- https://learn.microsoft.com/en-us/rest/api/azure/devops/build/builds/get-changes-between-builds?view=azure-devops-rest-7.1
- https://learn.microsoft.com/en-us/rest/api/azure/devops/build/builds/get-build-work-items-refs?view=azure-devops-rest-7.1

These calls return the ID of the WI/Commits, but not the state they were at the time the pipeline was run. So when you get the details via other API calls they will be the current details.

By default rerunning the a build (or classic release) will always get different associations as the API calls return the changes since the last successful release. However to get around this I have created a few options

- You can override the last successful build/release ID using the parameter `overrideBuildReleaseId`, you also override the current/active build/release ID `OverrideActiveBuildReleaseId` (thanks to a very recent community PR) - so these parameters allow to build a range. of builds to consider
- You can return the WI list based on a WI Query - check the [WIKI](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task) for `wiqlWhereClause `
- If you just want to rerun a build's release notes generation look at the [command line tester](https://github.com/rfennell/AzurePipelines/tree/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole). This was designed to speed up creation of handlebars templates, but can be used to re-run notes generation for builds.  Note: it is no so useful for Classic releases as these are very much point in time view of all the sub parts.  

Hope one of that helps

**Reply by khopithan** [reply]
Thank you 🙂
